nodejs/aws-infra/experimental/cluster.ts

```diff
@@ -0,0 +1,132 @@
+
+import * as mod from ".";
+
+import { Network } from "./../network";
+
+import * as utils from "../utils";
+
+/**
+ * A Cluster is a general purpose ECS cluster configured to run in a provided Network.
+ */
+export class Cluster extends pulumi.ComponentResource {
+    public readonly instance: aws.ecs.Cluster;
+
+    /**
+     * The network in which to create this cluster.
+     */
+    public readonly network: Network;
+    /**
+     * The ECS Cluster's Security Group.
+     */
+    public readonly instanceSecurityGroup: aws.ec2.SecurityGroup;
+
+    constructor(name: string, args: ClusterArgs, opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra:x:Cluster", name, {
+            ...args,
+        }, opts);
+
+        // First create an ECS cluster.
+        const parentOpts = { parent: this };
+        const instance = new aws.ecs.Cluster(name, args, parentOpts);
+
+        const network = args.network || Network.getDefault();
+
+        // Create the EC2 instance security group
+        const ALL = {
+            fromPort: 0,
+            toPort: 0,
+            protocol: "-1",  // all
+            cidrBlocks: [ "0.0.0.0/0" ],
+        };
+
+        // IDEA: Can we re-use the network's default security group instead of creating a specific
+        // new security group in the Cluster layer?  This may allow us to share a single Security Group
+        // across both instance and Lambda compute.
+        const instanceSecurityGroup = args.instanceSecurityGroup || new aws.ec2.SecurityGroup(name, {
+            vpcId: network.vpcId,
+            ingress: [
+                // Expose SSH
+                {
+                    fromPort: 22,
+                    toPort: 22,
+                    protocol: "TCP",
+                    cidrBlocks: [ "0.0.0.0/0" ],
+                },
+                // Expose ephemeral container ports to Internet.
+                // TODO: Limit to load balancer(s).
+                {
+                    fromPort: 0,
+                    toPort: 65535,
+                    protocol: "TCP",
+                    cidrBlocks: [ "0.0.0.0/0" ],
+                },
+            ],
+            egress: [ ALL ],  // See TerraformEgressNote
+            tags: { Name: name },
+        }, parentOpts);
+
+        this.instance = instance;
+        this.network = network;
+        this.instanceSecurityGroup = instanceSecurityGroup;
+
+        this.registerOutputs({
+            instance,
+            network,
+            instanceSecurityGroup,
+        });
+    }
+}
+
+(<any>Cluster).doNotCapture = true;
+
+// The shape we want for ClusterArgs.  We don't export this as 'Overwrite' types are not pleasant to
+// work with. However, they internally allow us to succinctly express the shape we're trying to
+// provide. Code later on will ensure these types are compatible.
+type OverwriteShape = utils.Overwrite<aws.ecs.ClusterArgs, {
+    network: Network;
+    instanceSecurityGroup?: aws.ec2.SecurityGroup;
+}>;
```

Comment:
implementation design:  We use the 'overwrite' shape internally to succinctly define the we want in terms of a change to an underlying args type.  we then expose a real interface with all the arg-type properties copied into it.  We then ensure these types are compatible below.